# Exploring eBay Car Sales Data

## Introduction

This project involves working with a dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website.

A user named Orgesleka scraped and uploaded the dataset to Kaggle originally. The original one cannot be found on Kaggle any longer, but it can be found here: https://data.world/data-society/used-cars-data

The dataset used in this project is a sample of the original one (50,000 data points from the full dataset). Below is the data dictionary:

| **Column name** | **Description** |
| :----- | :------ |
| **dateCrawled** | When this ad was first crawled. All field-values are taken from this date |
| **name** | Name of the car |
| **seller** | Whether the seller is private or a dealer |
| **offerType**| The type of listing |
| **price** | The price on the ad to sell the car |
| **abtest** | Whether the listing is included in an A/B test |
| **vehicleType** | The vehicle Type |
| **yearOfRegistration** | The year in which the car was first registered |
| **gearbox** | The transmission type |
| **powerPS** | The power of the car in PS |
| **model** | The car model name |
| **odometer** | How many kilometers the car has driven |
| **monthOfRegistration** | The month in which the car was first registered |
| **fuelType** | What type of fuel the car uses |
| **brand** | The brand of the car |
| **notRepairedDamage** | If the car has a damage which is not yet repaired |
| **dateCreated** |  The date on which the eBay listing was created |
| **nrOfPictures** | The number of pictures in the ad |
| **postalCode** | The postal code for the location of the vehicle |
| **lastSeenOnline** | When the crawler saw this ad last online |

The aim of this project is to clean the data and analyze the included used car listings.

## Import relevant libraries

In [1]:
import pandas as pd
import numpy as np

## Load the data

In [2]:
# Read the .csv file and save it to a variable
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
# Display the first 5 rows of the dataset 
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are 15 columns with the object dtype and 5 with int dtype. Some of the columns have NaN values which should be dealt with. The column names use the camel case instead of the snake case which is the preferred case for naming columns in Python. 

## Data Cleaning

### Cleaning the column names

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# Change the case of column names to snake case and reassign them to the 
# columns attribute of the dataframe
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

Changing the case and renaming some of the column names were done to suit the Python Programming convention.

### Initial Exploration and cleaning

In [8]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The `seller` and `offer_type` have mostly one value that are candidates to be dropped. The `date_crawled`, `ad_created` and `last_seen` columns contain datetime values but they are stored as string. Hence, they need preprocessing. The `price` and `odometer` columns are numeric values stored as text.

In [9]:
# Remove non-numeric characters from the `price` column
autos['price'] = autos['price'].str.replace("$", "")
autos['price'] = autos['price'].str.replace(",", "")
autos['price'] = autos['price'].astype(int)

# Remove non-numeric characters from the `odometer` column
autos['odometer'] = autos['odometer'].str.replace("km", "")
autos['odometer'] = autos['odometer'].str.replace(",", "")
autos['odometer'] = autos['odometer'].astype(int)

In [10]:
# Rename the `odometer` column to `odometer_km`
autos = autos.rename({"odometer":"odometer_km"}, axis=1)

In [11]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   

   power_PS   model  odometer_km  registration_month fuel_type    brand  \
0       158  andere       150000                   3       lpg  peugeot   
1       286     7er       150000                   6    benzin      bmw   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

## Exploring the `odometer_km` and `price` columns

### The `odometer_km` column

In [12]:
autos['odometer_km'].unique().shape

(13,)

In [13]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [14]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [15]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

### The `price` column

In [16]:
autos['price'].unique().shape

(2357,)

In [17]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [18]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

According to popular websites for online car sales such AutoScout24 and Mobile, the price range is approximately `$500` - `$100000`. This price range can then be used as a criteria for outliers removal.

In [19]:
# Removing outliers based on the price column
autos = autos[autos['price'].between(500, 100000)]

In [20]:
# Description after the outlier removal
autos.describe()

price  registration_year      power_PS    odometer_km  \
count  45058.000000       45058.000000  45058.000000   45058.000000   
mean    6177.407342        2005.063918    121.336788  125364.752097   
std     7636.932471          89.689852    205.622294   39542.959058   
min      500.000000        1000.000000      0.000000    5000.000000   
25%     1500.000000        2000.000000     75.000000  100000.000000   
50%     3500.000000        2004.000000    110.000000  150000.000000   
75%     7900.000000        2008.000000    150.000000  150000.000000   
max    99900.000000        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        45058.000000         45058.0  45058.000000  
mean             5.907675             0.0  51283.869169  
std              3.633063             0.0  25700.705805  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30952.000000  
50%              6.000000             0.0  50226.000000  
75%              9.000000             0.0  72156.750000  
max             12.000000             0.0  99998.000000

## Exploring the date columns

As stated early on, the `date_crawled`, `ad_created` and `last_seen` columns are all identified as string values by pandas. Because these three columns are represented as strings, it is necessary to convert the data into a numerical representation so they can be understood quantitatively.

In [21]:
# Visualize the columns to be considered
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

### "date_crawled" column

In [22]:
# Get the distribution for the data in the `date_crawled` column
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64

The cars were crawled from **2016-03-05** to **2016-04-07**. **2016-04-07** is the day with the least number of ad crawls (61 - approx. `0.1%`) while the most of number ad crawls is **2016-04-03** (1749 - approx. `3.9%`).

### "ad_created" column

In [23]:
# Get the distribution for the data in the `ad_created` column
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, Length: 76, dtype: float64

### "last_seen" column

In [24]:
# Get the distribution for the data in the `last_seen` column
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen, dtype: float64

The most of the ads were taken off the website on **2016-04-06** (`22%`).

### "registration_year" column

In [25]:
# The distribution of the `registration_year` column
autos["registration_year"].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum value is `1000`, the year before cars were invented. The maximum value is `9999` which several years in the future. 

### Dealing with Incorrect Registration Year Data

A car can't be registered after an ad of it has been posted on the website. Hence, any vehicle with a registration year above **2016** is inaccurate. For the lower bound, it could be somewhere in the first few decades of the **1900s**.

In [26]:
# Check the number of listings with cars that fall outside 
# the 1900 - 2016 interval 
autos[autos["registration_year"].between(1900, 2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 1

In [27]:
autos["registration_year"].value_counts().sort_index().head()

1000    1
1001    1
1910    2
1927    1
1929    1
Name: registration_year, dtype: int64

Looking at the first 5 values, the realistic year for registration starts from **1910**. Therefore, it fits the purpose of using it as the lower bound to remove the outliers. Besides, there's still a significant number of values when a subset of the data is selected based on the interval, 1900 - 2016.

In [28]:
# Remove the outliers
autos = autos[autos["registration_year"].between(1900, 2016)]

In [29]:
# Calculate the distribution of the remaining values
autos["registration_year"].value_counts(normalize=True)

2005    0.066168
2000    0.062702
2004    0.061986
2006    0.061593
2003    0.061385
          ...   
1948    0.000023
1938    0.000023
1939    0.000023
1943    0.000023
1952    0.000023
Name: registration_year, Length: 78, dtype: float64

Most of the cars are registered in the 2000s.

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. Aggregation can be used to understand the `brand` column.

In [30]:
# Check the distribution of the `brand` column
autos["brand"].value_counts()

volkswagen        9180
bmw               4980
mercedes_benz     4427
opel              4347
audi              3920
ford              2822
renault           1902
peugeot           1272
fiat              1019
seat               775
skoda              743
smart              655
mazda              650
nissan             650
citroen            613
toyota             581
hyundai            440
sonstige_autos     415
mini               406
volvo              406
honda              338
mitsubishi         335
kia                318
alfa_romeo         287
chevrolet          261
suzuki             256
porsche            252
chrysler           156
dacia              123
jeep               106
land_rover          98
daihatsu            94
subaru              86
saab                72
jaguar              70
daewoo              58
rover               54
trabant             48
lancia              43
lada                26
Name: brand, dtype: int64

The **Volkswagen** has the highest number of cars in the data, followed by **BMW**, **Mercedes Benz**, **Opel**, **Audi** and **Ford** in that order. The data is skewed to the top 7 brands.

In [31]:
# Select the index of the top 7 brands
top7_brands = autos["brand"].value_counts().head(7).index

In [32]:
# Get the mean prices for the top 7 brands
mean_prices_top7_brands = {}

for brand in top7_brands:
    mean_price = autos["price"].loc[autos["brand"] == brand].mean()
    mean_prices_top7_brands[brand] = mean_price
    
print(mean_prices_top7_brands)

{'volkswagen': 5783.622984749455, 'bmw': 8447.069879518072, 'mercedes_benz': 8666.677208041563, 'opel': 3394.0395675178283, 'audi': 9571.457397959184, 'ford': 4247.120481927711, 'renault': 2819.059411146162}


The most expensive among the top-7 brands is `Audi` -  approx. **9571**. The next expensive one is `Mercedes Benz` with a mean price of approx. **8666**, followed by `BMW` (approx. **8447**).The cheapest among the top-7 brands is `Renault` (approx. **2819**).

## Storing Aggregate Data in a DataFrame

In [33]:
# Get the mean mileage for the top 7 brands
mean_mileage_top7_brands = {}

for brand in top7_brands:
    mean_mileage = autos["odometer_km"].loc[autos["brand"] == brand].mean()
    mean_mileage_top7_brands[brand] = mean_mileage
    
print(mean_mileage_top7_brands)

{'volkswagen': 128234.74945533769, 'bmw': 132928.71485943775, 'mercedes_benz': 131083.12627061215, 'opel': 128012.42236024844, 'audi': 128941.32653061225, 'ford': 123520.55279943303, 'renault': 126351.20925341746}


In [34]:
# Create a series for the mean prices
mean_prices_series = pd.Series(mean_prices_top7_brands)
mean_prices_series

volkswagen       5783.622985
bmw              8447.069880
mercedes_benz    8666.677208
opel             3394.039568
audi             9571.457398
ford             4247.120482
renault          2819.059411
dtype: float64

In [35]:
# Create a series for the mean mileage
mean_mileage_series = pd.Series(mean_mileage_top7_brands)
mean_mileage_series

volkswagen       128234.749455
bmw              132928.714859
mercedes_benz    131083.126271
opel             128012.422360
audi             128941.326531
ford             123520.552799
renault          126351.209253
dtype: float64

In [36]:
# Create a dataframe using the mean price series
df = pd.DataFrame(mean_prices_series, columns=["mean_price"])

# Add the mean mileage series to the dataframe
df["mean_mileage"] = mean_mileage_series
df

mean_price   mean_mileage
volkswagen     5783.622985  128234.749455
bmw            8447.069880  132928.714859
mercedes_benz  8666.677208  131083.126271
opel           3394.039568  128012.422360
audi           9571.457398  128941.326531
ford           4247.120482  123520.552799
renault        2819.059411  126351.209253

From the results, the top-5 brands have high mileages. Yet, their prices are still among the highest, an indication that the mileage has less impact on the pricing.

## Conclusion

In this project, a sample (50,000 data points) of the dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website was explored and analyzed.

The sample dataset was cleaned to facilitate the accurate analysis. From the exploration, the dataset is based in 2016. It was discovered that the top-7 brands in the car listings are **Volkswagen**, **BMW**, **Mercedes Benz**, **Opel**, **Audi**, **Ford** and **Renault**. Among these brands, `Audi` is the most expensive followed by `Mercedes Benz` and `BMW` while `Renault` is the cheapest. It was also recognised that despite the high mileage covered by these top brands, it had no impact on the pricing; they are still the expensive ones.